### DISCUSSION TITLE
Are in-order acknowledgments a requirement?

### DISCUSSION THREAD

**Discussion by fracasula** [emqx/emqx#4703] [Ideas]
Hello guys :wave: I have a simple question for you.

The MQTT v5 specs say that acknowledgement packets must be sent in the order in which the corresponding `PUBLISH` packets were received.

> The Client MUST send PUBACK packets in the order in which the corresponding PUBLISH packets were received (QoS 1 messages) [MQTT-4.6.0-2]
> The Client MUST send PUBREC packets in the order in which the corresponding PUBLISH packets were received (QoS 2 messages) [MQTT-4.6.0-3]
> The Client MUST send PUBREL packets in the order in which the corresponding PUBREC packets were received (QoS 2 messages) [MQTT-4.6.0-4]

https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901240

I checked with the [VerneMQ](https://github.com/vernemq/vernemq) guys and that it is not a requirement for their broker.

Is it a requirement when using the EMQX Broker instead?

Thanks!

**Comment by tigercl** [comment]
@fracasula In fact, we currently have no mandatory requirements for this. But we are considering whether to strictly abide by the MQTT spec. What are your thoughts on this?

**Comment by fracasula** [comment]
> @fracasula In fact, we currently have no mandatory requirements for this. But we are considering whether to strictly abide by the MQTT spec. What are your thoughts on this?

Hi @tigercl! I'm actually glad that you guys aren't relying on the acks to be sent back in order. I'm not sure I understand the reasoning behind those specs so I'm investigating a bit more to see if I can understand why it was specced out that way.

Btw, I think not having the requirement to send the acks in order strongly simplifies the client implementation and I can see use cases where the client can leverage that to increase concurrency while keeping in-order guarantees  per topic/clientID. I believe that is something that you're offering out of the box with the `hash_clientid` option for shared-subscriptions but not all brokers offer that.

Anyway, as I said, I'll try to gather more information about this and possibly post an update here before closing if that is OK with you. In the meantime, thanks for getting back to me :+1:

**Comment by tigercl** [comment]
@fracasula I'm glad you mentioned its role in improving concurrency. It provides us with a new perspective to think about this matter.
